Mathlib/Combinatorics/SimpleGraph/EdgeLabelling.lean

```diff
@@ -0,0 +1,227 @@
+edge-colouring, but we reserve that terminology for labellings where incident edges cannot share a
+label.
+-/
+def EdgeLabelling (G : SimpleGraph V) (K : Type*) :=
+  G.edgeSet → K
+
+instance [DecidableEq V] [Fintype G.edgeSet] [Fintype K] : Fintype (EdgeLabelling G K) :=
+  Pi.instFintype
+
+instance [Nonempty K] : Nonempty (EdgeLabelling G K) :=
+  Pi.instNonempty
+
+instance [Inhabited K] : Inhabited (EdgeLabelling G K) :=
+  Pi.instInhabited
+
+instance [Subsingleton K] : Subsingleton (EdgeLabelling G K) :=
+  Pi.instSubsingleton
+
+instance [Unique K] : Unique (EdgeLabelling G K) :=
+  Pi.unique
+
+/--
+An edge labelling of the complete graph on `V` with labels in type `K`. Sometimes this is called an
+edge-colouring, but we reserve that terminology for labellings where incident edges cannot share a
+label.
+-/
+abbrev TopEdgeLabelling (V K : Type*) :=
+  EdgeLabelling (⊤ : SimpleGraph V) K
+
+theorem card_topEdgeLabelling [DecidableEq V] [Fintype V] [Fintype K] :
+    card (TopEdgeLabelling V K) = card K ^ (card V).choose 2 :=
+  Fintype.card_fun.trans (by rw [← edgeFinset_card, card_edgeFinset_top_eq_card_choose_two])
+
+namespace EdgeLabelling
+
+/--
+Convenience function to get the colour of the edge `x ~ y` in the colouring of the complete graph
+on `V`.
+-/
+def get (C : EdgeLabelling G K) (x y : V) (h : G.Adj x y) : K :=
+  C ⟨s(x, y), h⟩
+
+lemma get_eq (C : EdgeLabelling G K) (x y : V) (h : G.Adj x y) : C.get x y h = C ⟨s(x, y), h⟩ :=
+  rfl
+
+variable {C : EdgeLabelling G K}
+
+theorem get_comm (x y : V) (h) : C.get y x h = C.get x y h.symm := by
+  simp [EdgeLabelling.get, Sym2.eq_swap]
+
+@[ext]
+theorem ext_get {C' : EdgeLabelling G K}
+    (h : ∀ x y, (h : G.Adj x y) → C.get x y h = C'.get x y h) : C = C' := by
+  refine funext fun ⟨e, he⟩ => ?_
+  induction e using Sym2.inductionOn
+  exact h _ _ he
+
+/-- Compose an edge-labelling with a function on the colour set. -/
+def compRight (C : EdgeLabelling G K) (f : K → K') : EdgeLabelling G K' :=
+  f ∘ C
+
+/-- Compose an edge-labelling with a graph embedding. -/
+def pullback (C : EdgeLabelling G K) (f : G' ↪g G) : EdgeLabelling G' K :=
+  C ∘ f.mapEdgeSet
+
+@[simp]
+theorem pullback_apply {f : G' ↪g G} e : C.pullback f e = C (f.mapEdgeSet e) :=
+  rfl
+
+@[simp]
+theorem get_pullback {f : G' ↪g G} (x y) (h : G'.Adj x y) :
+    (C.pullback f).get x y h = C.get (f x) (f y) (by simpa) :=
+  rfl
+
+@[simp]
+theorem compRight_apply (f : K → K') (e) : C.compRight f e = f (C e) :=
+  rfl
+
+@[simp]
+theorem compRight_get (f : K → K') (x y) (h : G.Adj x y) :
+    (C.compRight f).get x y h = f (C.get x y h) :=
+  rfl
+
+attribute [scoped instance] Sym2.Rel.setoid in
+/-- Construct an edge labelling from a symmetric function taking values everywhere except the
+diagonal.
+-/
+def mk (f : ∀ x y : V, G.Adj x y → K)
+    (f_symm : ∀ (x y : V) (H : G.Adj x y), f y x H.symm = f x y H) : EdgeLabelling G K :=
+  fun ⟨e, he⟩ => by
```

Comment:
This one was written before the code generator supported Subtype.recOn, do you know if that's still the case? (If this question doesn't make sense, let me know and I'll investigate myself!)

Reply:
I'm sorry, I don't know what you mean.